File: 235-LowestCommonAncestorOfABinarySearchTree/235-LowestCommonAncestorOfABinarySearchTree.cpp
// Last updated: 6/11/2026, 11:34:24 AM
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* helper(TreeNode* root, TreeNode* p, TreeNode* q){
        if(root==nullptr){
            return nullptr;
        }
        if(root==p || root==q){
            return root;
        } 

        TreeNode* sn1 = helper(root->left, p, q);
        TreeNode* sn2 = helper(root->right, p, q);

        if(sn1 != nullptr && sn2!=nullptr){
            return root;
        }
        if(sn1!=nullptr){
            return sn1;
        }
        if(sn2!=nullptr){
            return sn2;
        }
        return nullptr;
    }

    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        return helper(root, p, q);
    }
};

File: 235-LowestCommonAncestorOfABinarySearchTree/235-LowestCommonAncestorOfABinarySearchTree.py
# Last updated: 6/11/2026, 11:34:30 AM
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        # if(root is None):
        #     return None
        # if(root.val == p.val or root.val == q.val):  #mistake:- not knowing p and q are treenodes and not values
        #     return root
        # left = self.lowestCommonAncestor(root.left, p ,q)
        # right = self.lowestCommonAncestor(root.right, p, q)
        # if(right is None):
        #     return left
        # if(left is None):
        #     return right
        # if(left and right):
        #     return root
        # if(left is None and right is None):
        #     return None
        # print(left,right)
        if(root is None):
            return None
        cur = root
        while(cur):
            if(p.val>cur.val and q.val>cur.val):
                cur = cur.right
            elif(p.val<cur.val and q.val<cur.val):
                cur = cur.left
            else:
                return cur

